### DISCUSSION TITLE
Enhancement: markdoc/next.js TypeScript emit

### DISCUSSION THREAD

**Discussion by fbartho** [markdoc/markdoc#89] [General]
I've been fiddling with a custom NextJS plugin (reason: I have differing opinions™  than NextJS does about how to structure my project, not your problem).

In this project I decided to use the code in https://github.com/markdoc/next.js/blob/main/src/loader.js as an example on the right way to integrate -- and it works! Great!!

In my project I'm using TypeScript, so I set about finding the right types to use, and I found myself confounded to learn that it wasn't possible. The types of `Markdoc.transform`'s `Config` just don't line up with the values that are created in `loader.js`.

It *works* so clearly this is some sort of TypeScript type mismatch.

The exact error I get is:
`Type '({ id, level, children, className }: HeadingProps) => Element' is not assignable to type 'string'` on the `nodes` parameter

```typescript
import * as config from "./schema/config";
import * as tags from "./schema/tags";
import * as nodes from "./schema/nodes";
import * as functions from "./schema/functions";

import type { transform } from "@markdoc/markdoc";
type SchemaConfig = Exclude<Parameters<typeof transform>[1], undefined>;

export const defaultSchema: SchemaConfig = {
	tags,
	nodes, // Error occurs here
	functions,
	...config,
};
```

Some questions:
- Is there a better type I could use in this position?
- Are we worried about loader.js/runtime.js / the NextJS plugin getting out of sync with Markdoc's core? -- Should we convert the whole webpack-loader into TypeScript to at least get TypeScript compile-time validation?
- Could (should?) markdoc/NextJS emit .tsx files?

**Comment by fbartho** [comment]
It's very possible even very likely that I've confused myself -- the code I was lifting was generated by this segment of the loader: https://github.com/markdoc/next.js/blob/bb6033df2326521059ee5d93adee8330daca0cac/src/loader.js#L147-L160

What's the right type for `const schema: ?`

**Comment by mfix-stripe** [comment]
Hey @fbartho! Sorry for the delay here. 

There is a lot to unpack here, but I will try my best 🙂 

FWIW, the loader in the Next.js plugin is very Next.js-specific, so might be extracting a generic one from it might not be the best place to start. For example, you could create a pretty basic webpack loader like this:

```js

   
const Markdoc = require('@markdoc/markdoc');

module.exports = function (source) {
  const {schema} = this.getOptions() || {};

  const ast = Markdoc.parse(source);
  const transformed = Markdoc.transform(ast, schema);
  const output = Markdoc.renderers.reactStatic(transformed);

  return `import React from 'react'; export default ${output}`;
};
```

As for the [`Config`](https://github.com/markdoc/markdoc/blob/main/src/types.ts#L19) mismatches, it might be a couple things. `MarkdocNextJsConfig` is the `Config` type that we use in `loader.js`: https://github.com/markdoc/next.js/blob/main/src/index.d.ts#L14


> Is there a better type I could use in this position?

I think you want to use [`Config`](https://github.com/markdoc/markdoc/blob/main/src/types.ts#L19)

> Are we worried about loader.js/runtime.js / the NextJS plugin getting out of sync with Markdoc's core? -- Should we convert the whole webpack-loader into TypeScript to at least get TypeScript compile-time validation?

I'm not worried about this getting out of date, since we have pretty good unit tests. However, I could see us getting advantages from TypeScript (for one, we don't need to manually create an index.d.ts file). Then again, this will require a bunch more build tooling in order to create and publish the library itself, where right now it is very simple—just export some JS files. 

> Could (should?) markdoc/NextJS emit .tsx files?

I'm not 100% sure what you mean here, so excuse me if I'm off. The loader itself just emits a string containing an exported React function. The emitted blob isn't actually a file, and therefore doesn't have an extension. 
However, it does require further transpilation, which is why we add `options.defaultLoaders.babel` here: https://github.com/markdoc/next.js/blob/main/src/index.js#L10

**Comment by mfix-stripe** [comment]
> What's the right type for `const schema: ?`

[`Config`](https://github.com/markdoc/markdoc/blob/main/src/types.ts#L19) 🙂

**Comment by fbartho** [comment]
Thanks for the answers, I'm happy to close this ticket at this point, unless you want to execute on changes from it -- Don't feel obligated to read further.

## Answering your question

>>Could (should?) markdoc/NextJS emit .tsx files?

>I'm not 100% sure what you mean here, so excuse me if I'm off.

What I meant there was that `loader.js` builds a source-file as a large string, that is interpreted by the other JS/JSX loaders -- I was curious if it should instead build a file that is first processed by the `.tsx` loader before being processed by the js loader. This is a little silly of a request, reflecting how I was proceeding: I basically exported the output of loader.js to a file, and then tried to add strong types to it.

## What am I doing
I'm actually not trying to make a webpack loader in my case _but I am still using NextJS_.

Instead, I built my own (very-lightweight) CMS -- which is then called from my `getStaticProps` method. -- This CMS then directly calls `Markdoc.parse/Markdoc.transform` -- almost exactly like loader.js does in the generated `getStaticProps` method it provides.

Advantages:
- I can strongly separate my content from my packages/code/components.
- I can easily patch/validate my frontmatter & inject or normalize data
- I can generate index/archive pages

<details><summary>Click to see the code!</summary>

```typescript
// [snipped imports & file-system paths]

class SiteDBRuntimeModule {
        // [snipped other methods that lookup the caches]
        
	async staticProps(file: SourceFileEntry): Promise<{ props: MarkdocComponentProps }> {
		// Significant chunks vendored from: https://github.com/markdoc/next.js/blob/main/src/loader.js

		const frontmatter = await this.frontmatter(file);
		const source = await this.updatedSource(file);
		const ast = Markdoc.parse(source);

		const logger = console;
		const errors = Markdoc.validate(ast)
			// tags are not yet registered, so ignore these errors
			.filter((e) => e.error.id !== "tag-undefined")
			.filter((e) => {
				switch (e.error.level) {
					case "debug":
					case "error":
					case "info": {
						logger[e.error.level](e.error.message);
						break;
					}
					case "warning": {
						logger.warn(e.error.message);
						break;
					}
					case "critical": {
						logger.error(e.error.message);
						break;
					}
					default: {
						logger.log(e.error.message);
						break;
					}
				}
				return e.error.level === "critical";
			})
			.flatMap((e) => {
				const lines = source.split("\n");

				const message = [e.error.message, ...lines.slice(...e.lines)];

				if (
					e.error &&
					e.error.location &&
					e.error.location.start &&
					(e.error.location.start as any).offset
				) {
					const prev = lines.slice(0, e.lines[0]).join("\n").length;
					const diff = (e.error.location.start as any).offset - prev;

					const pointer = `${" ".repeat(diff)}^`;
					message.push(pointer);
				}

				// add extra newline between errors
				message.push("");
				return message;
			});

		if (errors.length) {
			throw new Error(errors.join("\n"));
		}

		const rest = schemaSansComponents();

		const partialStrings = await gatherPartials(ast, path.posix.resolve(schemaDir, "partials"));

		// Ensure Node.transformChildren is available
		const partials = Object.entries(partialStrings).reduce((acc, [key, value]) => {
			acc[key] = Markdoc.parse(value);
			return acc;
		}, {} as Record<string, Node>);

		const cfg = {
			...rest,
			variables: {
				...(rest ? rest.variables : {}),
				// user can't override this namespace
				markdoc: { frontmatter },
				// Allows users to eject from Markdoc rendering and pass in dynamic variables via getServerSideProps
				// ...(context.variables || {}),
			},
			partials,
			source,
		};
		/**
		 * transform must be called in getStaticProps to support server-side rendering while
		 * accessing variables on the server
		 */
		const content = Markdoc.transform(ast, cfg);
		return {
			// Removes undefined
			props: JSON.parse(
				JSON.stringify({
					markdoc: {
						content,
						frontmatter,
						file: {
							path: file.relPath,
						},
					},
				}),
			),
		};
	}
}

/** Gathers the *referenced* partials to strings for later parsing */
async function gatherPartials(ast: Node, schemaDir: string) {
	let partials: Record<string, string> = {};

	for (const node of ast.walk()) {
		const file = node.attributes.file;

		if (
			node.type === "tag" &&
			node.tag === "partial" &&
			typeof file === "string" &&
			!partials[file]
		) {
			const filepath = path.posix.join(schemaDir, file);
			// parsing is not done here because then we have to serialize and reload from JSON at runtime
			const content = await fs.promises.readFile(filepath, { encoding: "utf8" });

			if (content) {
				const ast = Markdoc.parse(content);
				partials = {
					...partials,
					[file]: content,
					...(await gatherPartials(ast, schemaDir)),
				};
			}
		}
	}

	return partials;
}
```

</details>

**Comment by mfix-stripe** [comment]
Ah yes, makes sense. We want the loader to emit JSX-like contents for this particular plugin, as it has a dependency on Next.js, and therefore React. However, we don't want to introduce an explicit dependency on TypeScript, because folks should be able to use this loader with out it. 

As for calling `Markdoc.transform` yourself, you can organize your `config` however you prefer, and then just import it from somewhere in the file that contains `getStaticProps`.

> I can easily patch/validate my frontmatter & inject or normalize data

For what it is worth, you can create a custom [`document` node](https://markdoc.io/docs/nodes) and add a [custom validate function](https://markdoc.io/docs/validation) to handle the frontmatter validation in Markdoc itself.

@fbartho since this was more of a discussion in general, do you mind if I convert this issue into a [discussion](https://github.com/markdoc/markdoc/discussions)?

**Comment by mfix-stripe** [comment]
@fbartho converting this to a discussion—hope that is ok with you 🙂
